### ISSUE TITLE
Whatsup

### ISSUE THREAD

**Issue by cyberb** [cyberb/uapps#4]

**Comment by bittner** [comment]
Is this about a WhatsApp client on Ubuntu Phone? Something like [Loqui](https://support.mozilla.org/en-US/kb/whatsapp-alternatives-firefox-os#w_whatsapp-client-loqui), or a dedicated implementation, maybe even endorsed by WhatsApp/Facebook?

Probably [related discussion](https://lists.launchpad.net/ubuntu-phone/msg17241.html) on the ubuntu-phone mailing list.

**Comment by cyberb** [comment]
Yes this is WhatsApp app for Ubuntu Phone, I think it is fine to use existing code. As long as it is open source and available for installation as an app.

**Comment by dobey** [comment]
This is not feasible, without convincing WhatsApp to either port their own app, or open up the protocol to allow third party apps. Currently they will actively block third party clients, and ban the accounts of the users they find using them.

**Comment by cyberb** [comment]
OK then, we can ask their price for the port. Be it open source or closed source.

**Comment by bittner** [comment]
@cyberb I think you have to start with a proper, detailed description of your issue. It's very hard to give an estimate of the required effort for a user story with no content.

@dobey The [Loqui instant messenger](https://github.com/loqui/im) covers WhatsApp, and it does so with their own protocol as for what I remember (i.e. they don't distinguish any traffic coming from Loqui or their own app). In the long run, of course, it would probably be good to officially ask their permission. @Gioyik has worked on a port to Ubuntu Touch in [issue #434](https://github.com/loqui/im/issues/434).

**Comment by dobey** [comment]
@bittner Loqui seems to just be using their web interface, and not implementing any actual protocol as a new client. If it works and they haven't banned it or the people using it yet, then it only makes sense that it is using their web interface (especially since it's a thing in Firefox). The last comment on that referenced issue is also from 6 months ago. WhatsApp has a web interface, but requires a phone with the official app, to be able to create an account and log in, IIRC.

**Comment by bittner** [comment]
@dobey That's a guess. I'd suggest you try the app to prove that your assumption is true.

Some code in the [huge coseme.js file](https://github.com/loqui/im/blob/dev/src/scripts/mozillahispano/coseme.js#L3396) suggests, more likely, they are communicating with a REST-like interface.

**Comment by dobey** [comment]
@bittner Like I said, it's using the WhatsApp web interface. What did you think REST is? That is very different from the actual WhatsApp client applications on iOS/Android, and is still lacking many important features of the service. It's not sufficient for providing a meaningful messaging app experience in the realm of Ubuntu's security model though, I think.

**Comment by Gioyik** [comment]
> This is not feasible, without convincing WhatsApp to either port their own app, or open up the protocol to allow third party apps.

Don't think this is going to happen any time soon.

> Loqui seems to just be using their web interface

Not sure what you mean with this, but we added in Loqui the support for Whatsapp a lot of time before the official web interface was launched. 

> not implementing any actual protocol as a new client

How to implement a new protocol of something you don't know? is this more useless work, isn't it?

> it only makes sense that it is using their web interface

Still not the point about how work the library that Loqui uses to connect with Whatsapp service.

> That is very different from the actual WhatsApp client applications on iOS/Android

Loqui and other apps that uses CoSeMe are very close about how they work on iOS/Android.

> still lacking many important features of the service

If I am not wrong the unique feature missing of the service is Voice Messages, and it's not covered yet because people who maintain CoSeMe have more things to do in their lives and they keep the rest of the features alive everything Whatsapp changes something in their connection apis. So, if you consider CoSeMe for your client, you will have most of the features ready to be implemented in the client using the lib.

Seems that is not clear how Loqui connects to Whatsapp. Loqui uses CoSeMe, a library that lets you connect to Whatsapp servers using an api who was discovered using reverse engineering and a valid token to get **green** access. Loqui uses CoSeMe to send requests and receive data directly from Whatsapp servers. CoSeMe is independent from official Whatsapp Web client, it doesn't need bridges or something special to work, is what it is, a lib to make request and get data from official servers.

I tried to port Loqui to Ubuntu Touch but it was difficult due to the lack of documentation and the SDK was not friendly to use as a development environment. Never run the emulator correctly for me.

**Comment by bittner** [comment]
@Gioyik What is the state of the Ubuntu Touch port today? Unchanged?  Would you be willing to try to make this work in case you're sponsored (with a Ubuntu Touch device, with some money raised via e.g. freedomsponsors.org) and you're getting some pair-programming support?

I think when we use the Ubuntu phone IRC channels, mainly `#ubuntu-app-devel` on freenode, we can get the fast responses we need and hints to documentation that is hard to find. Also, oxide (the web browser implementation on Ubuntu Touch) has improved, that could make things easier today.

@cyberb Could you update the [Whatsup](https://freedomsponsors.org/issue/751/whatsup) issue on FreedomSponsors? It should say "WhatsApp" and have a complete description, maybe even mentioning LoquiIM as a possible solution (and linking to its repo or website).

Afterwards we should reach out to the Ubuntu community and promote this crowd funding attempt, so that we can leverage the interest in a working WhatsApp implementation on Ubuntu Touch. It's probably the most requested app on the platform (I guess).

**Comment by cmeerw** [comment]
Hi there, @Gioyik has just pointed me to this issue here...

Anyway, I just had a quick look yesterday to see how much work it would be to get Loqui working in Chrome and it was surprisingly straightforward, just a few Javascript constructs to change. Well, except for a few FirefoxOS specific APIs, like raw TCP sockets.

So assuming that the rendering engine on Ubuntu phone is similar to Chrome, the GUI shouldn't be much of a problem. The main problem would be the ability to open a raw TCP socket - is there any API to do this on Ubuntu phone? Loqui currently uses https://developer.mozilla.org/en-US/docs/Mozilla/B2G_OS/API/TCPSocket on FirefoxOS (and Chrome would support https://developer.chrome.com/apps/sockets_tcp )

**Comment by cmeerw** [comment]
Just to provide an update, I have now added a new build target for LoquiIM to build a Chrome app and basic WhatsApp text messages seem to work reasonable well now. But as I said, the main issue is that we need support for raw TCP sockets and I don't know if Ubuntu phone does support any API for that.

**Comment by cmeerw** [comment]
I have now also updated the packaging for Ubuntu in the Loqui build and I get something I can run with ubuntu-html5-app-launcher in Ubuntu - it's similar to the Chrome version except that there is no raw TCP socket support, so the only thing we can connect to is an XMPP server via WebSockets.

**Comment by bittner** [comment]
@cmeerw Is there any way "normal users" can try that on their Ubuntu phones? Can you provide a click package and [upload it to the Ubuntu Store](https://myapps.developer.ubuntu.com/dev/click-apps/)?

**Comment by cmeerw** [comment]
@bittner at this stage it is premature to add it to the Ubuntu Store... But I can provider a web hosted version that you should then be able to open in the web browser on your device (as there are no additional APIs that I know of that click packages would be able to use, it shouldn't make a difference so far)...

**Comment by cmeerw** [comment]
@bittner a web hosted version of LoquiIM is now available at http://hosted.loqui.im - but it can only connect to XMPP servers via a WebSocket server due to lack of support for opening raw TCP sockets.

**Comment by bittner** [comment]
@cmeerw, that user interface looks (and feels) awesome!

Unfortunately, registering for WhatsApp doesn't work neither on Ubuntu Touch nor on the desktop (Ubuntu 14.04.4 LTS, both latest Chromium and Firefox). The show-stoppers are different though: On Ubuntu Touch it's not even possible to use the UI to submit a phone number, on the desktop entering a phone number is possible but no phone call or SMS ever arrives.

**Comment by cmeerw** [comment]
@bittner Just a guess, but the rendering engine in Ubuntu 14.04.4 could be too old for the UI to work properly - ubuntu-html5-app-launcher seems to work fine for the UI on my Ubuntu 16.04 machine.

Anyway, when trying to use the hosted app for WhatsApp registration, there are 2 issues I can see: the browser doesn't allow the app to set the "User-Agent" header for the HTTP request and the HTTP request gets rejected anyway because of CORS issues (the CORS issue would go away for an installed app, but the inability to set the "User-Agent" header still appears to apply to an installed app as well). And if we could get registration working, we would then run into missing support for raw TCP sockets...

**Comment by FaraoneLele** [comment]
@Gioyik  The problem that you had with the 'SDKdi Ubuntu and its general malfunction has been resolved a few weeks ago, it is now possible to use the mouse and other normally operating bugs have been fixed. Now maybe you could try again and succeed (though you had tried two years ago and now has changed a lot since then)
